hydroflow/src/builder/surface/map_scan.rs

```diff
@@ -0,0 +1,167 @@
+
+    fn into_parts(self, ctx: &mut Context) -> (Self::InputHandoffs, Self::Build) {
+        let state_handle = ctx.add_state(RefCell::new(self.state));
+
+        let (connect, build) = self.prev.into_parts(ctx);
+        let build = MapPullBuild::new(build, wrap_func(self.func, state_handle));
+        (connect, build)
+    }
+}
+impl<Prev, Func, State, Out> AssembleFlowGraph for MapScanSurface<Prev, Func, State>
+where
+    Prev: PullSurface + AssembleFlowGraph,
+    Func: FnMut(&mut State, Prev::ItemOut) -> Out,
+    State: Any,
+{
+    fn insert_dep(&self, e: &mut super::FlowGraph) -> NodeId {
+        let my_id = e.add_node("Map");
+        let prev_id = self.prev.insert_dep(e);
+        e.add_edge((prev_id, my_id));
+        my_id
+    }
+}
+
+impl<Prev, Func, State, Out> PushSurface for MapScanSurface<Prev, Func, State>
+where
+    Prev: PushSurface,
+    Func: FnMut(&mut State, Prev::ItemOut) -> Out,
+    State: Any,
+{
+    type Output<Next> = Prev::Output<MapScanPushSurfaceReversed<Next, Func, State, Prev::ItemOut>>
+    where
+        Next: PushSurfaceReversed<ItemIn = Self::ItemOut>;
+
+    fn push_to<Next>(self, next: Next) -> Self::Output<Next>
+    where
+        Next: PushSurfaceReversed<ItemIn = Self::ItemOut>,
+    {
+        self.prev
+            .push_to(MapScanPushSurfaceReversed::new(next, self.func, self.state))
+    }
+}
+
+pub struct MapScanPushSurfaceReversed<Next, Func, State, In>
+where
+    Next: PushSurfaceReversed,
+    Func: FnMut(&mut State, In) -> Next::ItemIn,
+    State: Any,
+{
+    next: Next,
+    func: Func,
+    state: State,
+    _phantom: PhantomData<fn(In)>,
+}
+impl<Next, Func, State, In> MapScanPushSurfaceReversed<Next, Func, State, In>
+where
+    Next: PushSurfaceReversed,
+    Func: FnMut(&mut State, In) -> Next::ItemIn,
+    State: Any,
+{
+    pub fn new(next: Next, func: Func, state: State) -> Self {
+        Self {
+            next,
+            func,
+            state,
+            _phantom: PhantomData,
+        }
+    }
+}
+impl<Next, Func, State, In> AssembleFlowGraph for MapScanPushSurfaceReversed<Next, Func, State, In>
+where
+    Next: PushSurfaceReversed + AssembleFlowGraph,
+    Func: FnMut(&mut State, In) -> Next::ItemIn,
+    State: Any,
+{
+    fn insert_dep(&self, e: &mut super::FlowGraph) -> NodeId {
+        let my_id = e.add_node("Map");
+        let next_id = self.next.insert_dep(e);
+        e.add_edge((my_id, next_id));
+        my_id
+    }
+}
+
+impl<Next, Func, State, In> PushSurfaceReversed
+    for MapScanPushSurfaceReversed<Next, Func, State, In>
+where
+    Next: PushSurfaceReversed,
+    Func: FnMut(&mut State, In) -> Next::ItemIn,
+    State: Any,
+{
+    type ItemIn = In;
+
+    type OutputHandoffs = Next::OutputHandoffs;
+    type Build = MapPushBuild<Next::Build, MapScanFunc<Func, State, In, Next::ItemIn>, In>;
+
+    fn into_parts(self, ctx: &mut Context) -> (Self::OutputHandoffs, Self::Build) {
```

Comment:
nit: consider giving this a different name to imply it's doing something besides splitting up this thing into pieces, maybe "construct" or something? Not sure. Maybe fine as-is.